### PR TITLE
Improve checkpoint docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp #7666

Changes made here:
1. `saving and loading weights` -> `saving and loading checkpoints`
2. Moved `checkpoint io` to `saving and loading checkpoints`
3. Added new structure after discussing it with @kaushikb11 
```
Saving and Loading (general APIs)
Conditional Checkpointing (ModelCheckpoint)
Customize Checkpointing (checkpoint IO)
```
## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud

### CODE REVIEW

File: docs/source/common/weights_loading.rst
Comment (justusschock) [inline docs/source/common/weights_loading.rst:10]: if you call it checkpoints, this should also be reflected in the filename :)

File: docs/source/common/weights_loading.rst
Comment (justusschock) [inline docs/source/common/weights_loading.rst:13]:
```suggestion
Lightning provides functions to save and load checkpoints.
```

File: docs/source/common/weights_loading.rst
Comment (justusschock) [inline docs/source/common/weights_loading.rst:24]: Is the scaling factor not saved when using native amp?

**Review by justusschock — APPROVED** [review]

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst]: done! thanks :)

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst]: yeah for both. .updated!!

File: docs/source/common/checkpointing.rst
Comment (kaushikb11) [inline docs/source/common/checkpointing.rst:69]:
```suggestion
Manual Saving with Distributed Training Strategies
```

File: docs/source/common/checkpointing.rst
Comment (rohitgr7) [inline docs/source/common/checkpointing.rst:70]:
```suggestion
Manual Saving with Distributed Training Strategies
==================================================
```

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [inline docs/source/common/weights_loading.rst:7]: just fyi: I'll be renaming this file to `checkpointing.rst` once I get all the required approvals. Not doing it now for easy review for reviewers. Doc tests might fail because references are already updated at other places.

Open to better filename suggestions though :)

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1880]: This is the path to where the checkpoint gets saved, not the actual contents

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:105]: This form is less common, let's use the common term "Hyperparameters"?

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:88]:
```suggestion
Modifying Checkpoint When Saving and Loading
============================================
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:83]:
```suggestion
Not using :meth:`~pytorch_lightning.trainer.trainer.Trainer.save_checkpoint` can lead to unexpected behavior and potential deadlock. Using other saving functions will result in all devices attempting to save the checkpoint. As a result, we highly recommend using the Trainer's save functionality.
```
The majority of the docs and Lightning uses the American

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:84]:
```suggestion
If using custom saving functions cannot be avoided, we recommend using the :func:`~pytorch_lightning.utilities.distributed.rank_zero_only` decorator to ensure saving occurs only on the main process.
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:191]:
```suggestion
The :class:`~pytorch_lightning.callbacks.ModelCheckpoint` callback allows you to configure how & when checkpointing should happen. It follows the normal Callback hook structure so you can
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:222]:
```suggestion
- It gives you the ability to specify the ``dirpath`` and ``filename`` for your checkpoints. Filename can also be dynamic so you can inject the metrics that are being logged using :meth:`~pytorch_lightning.core.lightning.LightningModule.log`.
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:251]: K is not specified here. by default save_top_k=1
same in the example below.

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:265]:
```suggestion
- In case your training is interrupted for some reason, you can specify ``save_last=True`` to ensure that a checkpoint is being created at the end so that you can
  resume it or use it for :doc:`fault-tolerant training <../advanced/fault_tolerant_training>`.
```
That's not actually the use for it. 
last.ckpt is always getting saved when saving a regular checkpoint. It's an identical copy, with the key difference being that the name can always be determined as last.ckpt. This is normally not possible with save_last=False because the filename may contain arbitrary information. Fault-tolerant has it's own checkpoint file name and location, it's unrelated to this one.

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:267]:
```suggestion
- By default, the``ModelCheckpoint`` callback saves model weights, optimizer states, etc., but in case you have limited disk space or just need the model weights to be saved you can specify ``save_weights_only=True``.
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:268]:
```suggestion
- You can checkpoint at a regular time interval using ``train_time_interval`` argument independent of the steps or epochs.
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:272]:
```suggestion
The :class:`~pytorch_lightning.callbacks.ModelCheckpoint` callback is very robust and should cover 99% of the use-cases. If you find a use-case that is not configured yet, feel free to open an issue with a feature request on GitHub
```
word repetition: "case"

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:273]:
```suggestion
and the Lightning Team will be happy to integrate/help integrate it.
```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:311]:
```suggestion
Custom Checkpoint IO Plugin
===========================

```

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:317]: Can you remove the typing? It's noise for the reader and if they need that information, they would look it up in their IDE / API reference.

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:326]: let's remove the typing

File: docs/source/common/weights_loading.rst
Comment (awaelchli) [inline docs/source/common/weights_loading.rst:358]:
```suggestion
    Some ``TrainingTypePlugins`` like ``DeepSpeedPlugin`` do not support custom ``CheckpointIO`` as checkpointing logic is not modifiable.
```

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:265]: removed!
updated all!
thank you for extensive review 😃

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [inline docs/source/common/weights_loading.rst:191]: how & when --> when & which ?

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [inline docs/source/common/weights_loading.rst:95]: maybe this can be right after `Automatic Saving` section?

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [inline docs/source/common/weights_loading.rst:267]: probably a separate PR, but i'm thinking we can categorize this a bit more later e.g.

"when" = triggers (`every_n_epochs` / `every_n_training_steps` / `train_time_interval`)
"which" = `save_top_k`, `save_last`
"what" = `save_weights_only`
"where" = `dirpath`, `filename`, `auto_insert_metric_name`

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [inline docs/source/common/weights_loading.rst:266]: switch line 266 and 267 so that the "when" triggers can be together  (`every_n_epochs` / `every_n_training_steps` / `train_time_interval`)

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [inline docs/source/common/weights_loading.rst:289]: love the added distinction :)
(maybe "how the checkpoint is saved/loaded" --> "how the checkpoint is saved/loaded to storage") ?

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:267]: this is a newly added section. I'll improve it based on your points.

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:191]: how about?
```
The :class:`~pytorch_lightning.callbacks.ModelCheckpoint` callback allows you to configure how, when, and which checkpoints should be saved to storage.
```

File: docs/source/common/weights_loading.rst
Comment (ananthsub) [inline docs/source/common/weights_loading.rst:27]:
```suggestion
- Lightning Module state_dict
```

File: docs/source/common/weights_loading.rst
Comment (ananthsub) [inline docs/source/common/weights_loading.rst:53]: does this work? is `best_model_path` defined if there's no `monitor` set in the ModelCheckpoint callback?

File: docs/source/common/weights_loading.rst
Comment (ananthsub) [inline docs/source/common/weights_loading.rst:84]: For model parallel use cases, using `rank_zero_only` will not work. `rank_zero_only` is recommended only if all ranks hold the exact same state.

File: docs/source/common/weights_loading.rst
Comment (ananthsub) [inline docs/source/common/weights_loading.rst:19]: Checkpoint contents? This section is also useful/relevant for loading, not just saving

**Review by ananthsub — COMMENTED** [review]
thanks for working on this @rohitgr7 ! One observation is that there's a lot of text. Do you think we could add a diagram to explain the checkpoint saving and checkpoint loading flows? This could introduce readers to the components involved and when they take effect, and then they can read the docs for a deeper dive

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [reply docs/source/common/weights_loading.rst:191]: hm i guess "how" could encompass a lot, though probably avoid "how to storage" as that's very specifically IOPlugin

File: docs/source/common/weights_loading.rst
Comment (jjenniferdai) [reply docs/source/common/weights_loading.rst:53]: [yep](https://github.com/PyTorchLightning/pytorch-lightning/blob/3fcfd0214cc95fa7de49f6404421b5c6cd113c57/pytorch_lightning/callbacks/model_checkpoint.py#L677). I guess its the best 1/1, though the name is even more confusing when `save_top_k = -1` and `monitor = None`. maybe we could better rename this in the future, sth like `self.latest_model_path` ?

**Review by awaelchli — APPROVED** [review]

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:53]: updated!

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:84]: updated!

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst]: did some updates! mind review?

**Comment by rohitgr7** [comment]
> thanks for working on this @rohitgr7 ! One observation is that there's a lot of text. Do you think we could add a diagram to explain the checkpoint saving and checkpoint loading flows? This could introduce readers to the components involved and when they take effect, and then they can read the docs for a deeper dive

yes @ananthsub, that'd be really helpful. I'll be creating diagrams for hook flows. Added this to TODO. Will make for this one too later.

File: docs/source/index.rst
Comment (tchaton) [inline docs/source/index.rst:64]: Neat !

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:105]:
```suggestion
            saved. Otherwise, the optimizer states, lr-scheduler states, etc are added in the checkpoint too.
```

**Review by jjenniferdai — APPROVED** [review]
looks awesome :)

File: docs/source/common/weights_loading.rst
Comment (rohitgr7) [reply docs/source/common/weights_loading.rst:215]: done!

File: pytorch_lightning/core/saving.py
Comment (carmocca) [inline pytorch_lightning/core/saving.py:93]: Sphinx already includes the defaults

```suggestion
                returned by this module's state dict.
```

**Review by carmocca — APPROVED** [review]

**Pull request merged**
